docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+- Intentional multi-writer operation on tenants (multi-writer scenarios are assumed to be transient split-brain situations).
+- Sharding.
+
+## Impacted Components
+
+pageserver, control plane, safekeeper (optional)
+
+## Implementation Part 1: Correctness
+
+### Summary
+
+- **Generation numbers** are introduced for pageserver node lifetimes and tenant attachments.
+
+  - node generation increments each time a pageserver starts, before it can write to S3
+  - attachment generation increments each time the control plane modifies a tenant (`Project`)'s assigned pageserver (an _attachment_'s lifetime is the association between a tenant and a pageserver)
+  - the two generations are independent: a pageserver may restart while keeping the same
+    attachment generation numbers, and attachments may be changed while a pageserver's
+    generation number stays the same.
+
+- **Object keys are suffixed** with a tuple of attachment & node generation numbers
+- **Safety in split brain for multiple nodes running with
+  the same node ID** is provided by the pageserver node generation in the object key: the concurrent nodes
+  will not write to the same key.
+- **Safety for multiply-attached tenants** is provided by the
+  tenant attach generation in the object key: the competing pageservers will not
+  try to write to the same keys.
+- **Safety for deletions** is provided by pageservers validating with control plane that neither pageserver process nor the attachment has been superseded since it last updated a timeline's index.
+- **The control plane is used to issue generation numbers** to avoid the need for
+  a built-in consensus system in the pageserver, although this could in principle
+  be changed without changing the storage format.
+
+### Generation numbers
+
+Two logical entities will get new "generation numbers", which are monotonically increasing
+integers with a global guarantee that they will not be re-used:
+
+- Pageserver processes: for a given pageserver node ID, each time a pageserver process starts,
+  the per-node-ID generation number increases.
+- Tenant attachments: for a given tenant, each time its attachment is changed, a per-tenant generation
+  number increases.
+
+This provides some important invariants:
+
+- If there are multiple pageservers running with the same node ID, they are guaranteed to have
+  a different generation number.
+- If there are multiple pageservers running with the same node ID, we may unambiguously know which
+  of them "wins" by picking the higher generation number.
+- If two pageservers have the same tenant attached, they are guaranteed to have different attachment
+  generation numbers.
+
+The node generation number defines a global "in" definition for pageserver nodes: a node whose
+node generation matches the node generation the control plane most recently issued is a member of the cluster
+in good standing. Any node with any older generation number is not.
```

Comment:
Does this mean that when a pageserver restarts, it needs to communicate with the control plane to get a new node generation number?

That's not necessarily a problem, and we currently also download files files from S3 at pageserver startup. But would be nice if we could avoid it..

Reply:
It's unavoidable for there to be some synchronization at startup unfortunately, in the general case (i.e. running on less-trusted infra where it might e.g. run the same node twice by accident).  If/when we move to ephemeral IDs, this will be the case anyway: a pageserver won't have any work until it reports in to request some.

The impact is discussed in the "Availability" section later in the doc.